src/plonk/assigned.rs

```diff
@@ -0,0 +1,331 @@
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/0) - inv0(1/2) = 0 - 1/2 = -1/2
+        // This fails if subtraction is implemented using normal rules for rationals.
+        assert_eq!((b - a).evaluate(), (-a).evaluate());
+    }
+
+    #[test]
+    fn mul_rational_by_inv0_rational() {
+        // a = inv0(1/2)
+        // b = inv0(1/0)
+        let a = Assigned::Rational(Fp::one(), Fp::from(2));
+        let b = Assigned::Rational(Fp::one(), Fp::zero());
+
+        // inv0(1/2) * inv0(1/0) = 1/2 * 0 = 0
+        assert_eq!((a * b).evaluate(), Fp::zero());
+    }
+}
+
+#[cfg(test)]
+mod proptests {
+    use std::{
+        convert::TryFrom,
+        ops::{Add, Mul, Sub},
+    };
+
+    use pasta_curves::{arithmetic::FieldExt, Fp};
+    use proptest::{collection::vec, prelude::*, sample::select};
+
+    use super::Assigned;
+
+    #[derive(Clone, Debug)]
+    enum Operation {
+        Add,
+        Sub,
+        Mul,
+    }
+
+    const OPERATIONS: &[Operation] = &[Operation::Add, Operation::Sub, Operation::Mul];
+
+    impl Operation {
+        fn apply<F: Add<Output = F> + Sub<Output = F> + Mul<Output = F>>(&self, a: F, b: F) -> F {
+            match self {
+                Self::Add => a + b,
+                Self::Sub => a - b,
+                Self::Mul => a * b,
+            }
+        }
+    }
+
+    prop_compose! {
+        /// Use narrow that can be easily reduced.
+        fn arb_element()(val in any::<u64>()) -> Fp {
+            Fp::from(val)
+        }
+    }
+
+    prop_compose! {
+        fn arb_trivial()(element in arb_element()) -> Assigned<Fp> {
+            Assigned::Trivial(element)
+        }
+    }
+
+    prop_compose! {
+        /// Generates half of the denominators as zero to represent a deferred inversion.
+        fn arb_rational()(
+            numerator in arb_element(),
+            denominator in prop_oneof![Just(Fp::zero()), arb_element()],
+        ) -> Assigned<Fp> {
+            Assigned::Rational(numerator, denominator)
+        }
+    }
+
+    prop_compose! {
+        fn arb_testcase()(
+            num_operations in 1usize..5,
+        )(
+            values in vec(
+                prop_oneof![Just(Assigned::Zero), arb_trivial(), arb_rational()],
+                num_operations + 1),
+            operations in vec(select(OPERATIONS), num_operations),
+        ) -> (Vec<Assigned<Fp>>, Vec<Operation>) {
+            (values, operations)
+        }
+    }
+
+    proptest! {
+        #[test]
+        fn operation_commutativity((values, operations) in arb_testcase()) {
+            // Evaluate the values at the start.
+            let elements: Vec<_> = values.iter().cloned().map(|v| v.evaluate()).collect();
+
+            // Apply the operations to both the deferred and evaluated values.
+            let deferred_result = {
+                let mut ops = operations.iter();
+                values.into_iter().reduce(|a, b| ops.next().unwrap().apply(a, b)).unwrap()
```

Comment:
If I understand correctly, this is always doing something like `(((((v0 op0 v1) op1 v2) op2 v3) op3 v4) op4) v5` (or fewer operations), where each op is `+`, `-` or `*`. So this doesn't test `.invert()`.

Reply:
I extend this to unary opcodes in #417, which I left as a separate PR so this one didn't get too big.